demo.java

```diff
@@ -0,0 +1,166 @@
+    }); 
+         
+        final AmazonS3 s3Client = EventHandler.getS3Client(); 
+        //Create a new file in the Constants.SUMMARY_BUCKET 
+        logger.log("Map of statuses -> " + latestStatusForTrackingNumber); 
+        String summaryUpdateName = Long.toString(System.currentTimeMillis()); 
+         
+        EventHandler.getS3Client().putObject(Constants.SUMMARY_BUCKET, summaryUpdateName, latestStatusForTrackingNumber.toString());
+     
+        Waiter waiter = EventHandler.getS3Client().waiters().objectExists();
+        try{
+            waiter.run(new WaiterParameters<>(new GetObjectMetadataRequest(Constants.SUMMARY_BUCKET, summaryUpdateName)));
+            deleteProcessedFiles(filesToDelete);
+            logger.log("All updates successfully processed");
+        } catch (WaiterTimedOutException e){
+            throw new RuntimeException("Failed to write sumary status, will be retried in 15 minutes");
+        }
+    } 
+ 
+    private List<KeyVersion> processEventsInBucket(String bucketName, LambdaLogger logger, ConcurrentHashMap<String, Pair<Long, String>> latestStatusForTrackingNumber) { 
+ 
+        final AmazonS3 s3Client = EventHandler.getS3Client(); 
+        logger.log("Processing Bucket: " + bucketName); 
+ 
+        ObjectListing files = s3Client.listObjects(bucketName); 
+        List<KeyVersion> filesProcessed = new ArrayList<DeleteObjectsRequest.KeyVersion>(); 
+ 
+        for (Iterator<?> iterator = files.getObjectSummaries().iterator(); iterator.hasNext(); ) { 
+            S3ObjectSummary summary = (S3ObjectSummary) iterator.next(); 
+            logger.log("Reading Object: " + summary.getKey()); 
+ 
+            String trackingNumber = summary.getKey().split("--")[0]; 
+            Pair<Long, String> lastKnownStatus = latestStatusForTrackingNumber.get(trackingNumber); 
+ 
+            // Check if this shipment has already been delivered, skip this file 
+            if (lastKnownStatus != null && "DELIVERED".equals(lastKnownStatus.getRight())) { 
+                continue; 
+            } 
+ 
+            String fileContents = s3Client.getObjectAsString(bucketName, summary.getKey()); 
+ 
+            if (!isValidFile(fileContents)) { 
+                logger.log(String.format("Skipping invalid file %s", summary.getKey())); 
+                continue; 
+            } 
+             
+            if (!fileContents.contains("\n")) { 
+                 
+            } 
+            String[] lines = fileContents.split("\n"); 
+            String line1 = lines[0]; 
+            String line2 = lines[1]; 
+ 
+            String status = line1.split(":")[1]; 
+            Long timeStamp = Long.parseLong(line2.split(":")[1]); 
+ 
+ 
+            if (null == lastKnownStatus || lastKnownStatus.getLeft() < timeStamp) { 
+                lastKnownStatus = new MutablePair<Long, String>(timeStamp, status); 
+                latestStatusForTrackingNumber.put(trackingNumber, lastKnownStatus); 
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()` and `put()` may not be thread-safe at lines: **101, 104, 126, and 128**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).